onnxruntime/test/providers/qnn/qnn_basic_test.cc

```diff
@@ -0,0 +1,206 @@
+
+#include "test/common/tensor_op_test_utils.h"
+#include "test/framework/test_utils.h"
+#include "test/test_environment.h"
+#include "test/util/include/asserts.h"
+#include "test/util/include/default_providers.h"
+#include "test/util/include/inference_session_wrapper.h"
+#include "test/util/include/test_utils.h"
+
+#if !defined(ORT_MINIMAL_BUILD)
+// if this is a full build we need the provider test utils
+#include "test/optimizer/qdq_test_utils.h"
+#endif
+
+#include "gtest/gtest.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::logging;
+
+#define ORT_MODEL_FOLDER ORT_TSTR("testdata/")
+
+// in test_main.cc
+extern std::unique_ptr<Ort::Env> ort_env;
+
+namespace onnxruntime {
+namespace test {
+
+// test uses ONNX model so can't be run in a minimal build.
+// TODO: When we need QNN in a minimal build we should add an ORT format version of the model
+#if !defined(ORT_MINIMAL_BUILD)
+
+// Tests that the QNN EP is registered when added via the public C++ API.
+// Loads a simple ONNX model that adds floats.
+TEST(QnnEP, TestAddEpUsingPublicApi) {
+  {
+    // C++ API test
+    Ort::SessionOptions so;
+    onnxruntime::ProviderOptions options;
+
+#if defined(_WIN32)
+    options["backend_path"] = "QnnCpu.dll";
+#else
+    options["backend_path"] = "libQnnCpu.so";
+#endif
+
+    so.AppendExecutionProvider("QNN", options);
+
+    const ORTCHAR_T* ort_model_path = ORT_MODEL_FOLDER "constant_floats.onnx";
+    Ort::Session session(*ort_env, ort_model_path, so);
+
+    // Access the underlying InferenceSession.
+    const OrtSession* ort_session = session;
+    const InferenceSession* s = reinterpret_cast<const InferenceSession*>(ort_session);
+
+    bool have_qnn_ep = false;
+
+    for (const auto& provider : s->GetRegisteredProviderTypes()) {
+      if (provider == kQnnExecutionProvider) {
+        have_qnn_ep = true;
+        break;
+      }
+    }
+
+    ASSERT_TRUE(have_qnn_ep) << "QNN EP was not found in registered providers for session.";
+  }
+}
+
+#if defined(__aarch64__) || defined(_M_ARM64) || defined(__linux__)
+static void RunModelTest(
+    const GetQDQTestCaseFn& build_test_case,
+    const char* test_description,
+    const ProviderOptions& provider_options,
+    const EPVerificationParams& params = EPVerificationParams()) {
+  onnxruntime::Model model(test_description, false, DefaultLoggingManager().DefaultLogger());
+  Graph& graph = model.MainGraph();
+  ModelTestBuilder helper(graph);
+  build_test_case(helper);
+  helper.SetGraphOutputs();
+  ASSERT_STATUS_OK(model.MainGraph().Resolve());
+
+  // Serialize the model to a string.
+  std::string model_data;
+  model.ToProto().SerializeToString(&model_data);
+  RunAndVerifyOutputsWithEP(model_data, "QnnEP.TestQDQModel",
+                            DefaultQnnExecutionProviderWithOptions(provider_options),
+                            helper.feeds_, params);
+}
+
+// Check that QNN compiles DQ -> Conv -> Q as a single unit.
+TEST(QnnEP, TestQDQConvU8U8) {
```

Comment:
I adapted these tests from the xnnpack EP tests. They allow us to 1) test which nodes are assigned to an EP, and 2) check that the inference output matches with the CPU EP. Pretty nifty!